src/main/java/com/minecolonies/coremod/tileentities/TileEntityMultiBlock.java

```diff
@@ -0,0 +1,398 @@
+import net.minecraft.world.World;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.List;
+
+import static com.minecolonies.api.util.constant.Constants.*;
+import static com.minecolonies.api.util.constant.NbtTagConstants.*;
+import static com.minecolonies.coremod.util.SoundUtils.*;
+import static net.minecraft.util.EnumFacing.*;
+
+public class TileEntityMultiBlock extends TileEntity implements ITickable
+{
+    /**
+     * Max block range.
+     */
+    private static final int MAX_RANGE = 10;
+
+    /**
+     * Max block speed.
+     */
+    private static final int MAX_SPEED = 3;
+
+    /**
+     * Min block speed.
+     */
+    private static final int MIN_SPEED = 1;
+
+    /**
+     * Default gate and bridge range.
+     */
+    public static final int DEFAULT_RANGE         = 3;
+
+    /**
+     * Default gate and bridge range.
+     */
+    public static final int DEFAULT_SPEED        = 2;
+
+    /**
+     * The last redstone state which got in.
+     */
+    private boolean on = false;
+
+    /**
+     * The direction it should push or pull rom.
+     */
+    private EnumFacing direction = UP;
+
+    /**
+     * The output direction.
+     */
+    private EnumFacing output = DOWN;
+
+    /**
+     * The range it should pull to.
+     */
+    private int range = DEFAULT_RANGE;
+
+    /**
+     * The direction it is going to.
+     */
+    private EnumFacing currentDirection;
+
+    /**
+     * The progress it has made.
+     */
+    private int progress = 0;
+
+    /**
+     * Amount of ticks passed.
+     */
+    private int ticksPassed = 0;
+
+    /**
+     * Speed of the multiblock, max 3, min 1.
+     */
+    private int speed = 2;
+
+    /**
+     * Public constructor to create the tileEntity.
+     */
+    public TileEntityMultiBlock()
+    {
+        super();
+    }
+
+    /**
+     * Handle redstone input.
+     *
+     * @param signal true if positive.
+     */
+    public void handleRedstone(final boolean signal)
+    {
+        if(speed == 0)
+        {
+            speed = DEFAULT_SPEED;
+        }
+
+        if (signal != on && progress == range)
+        {
+            on = signal;
+            if (signal)
+            {
+                currentDirection = output;
+            }
+            else
+            {
+                currentDirection = direction;
+            }
+            progress = 0;
+        }
+    }
+
+
+    @Override
+    public void update()
+    {
+        if(world == null || world.isRemote)
+        {
+            return;
+        }
+        if (currentDirection == null && progress < range)
+        {
+            progress = range;
+        }
+
+        if(progress < range)
+        {
+            if (ticksPassed % ( TICKS_SECOND / speed) == 0)
+            {
+                handleTick();
+                ticksPassed = 1;
+            }
+            ticksPassed++;
+        }
+    }
+
+    /**
+     * Handle the tick, to finish the sliding.
+     */
+    public void handleTick()
```

Comment:
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') The Cyclomatic Complexity of this method "handleTick" is 18 which is greater than 15 authorized. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR') Refactor this method to reduce its Cognitive Complexity from 27 to the 20 allowed. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)